File: app/src/main/java/com/dertyp7214/appstore/settings/SettingsSwitch.java
/*
 * Copyright (c) 2018.
 * Created by Josua Lengwenath
 */

package com.dertyp7214.appstore.settings;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.SharedPreferences;

import static com.dertyp7214.appstore.Utils.getSettings;

public class SettingsSwitch extends Settings {

    private boolean checked;
    private CheckedChangeListener checkedChangeListener;

    public SettingsSwitch(String name, String text, Context context, boolean checked) {
        super(name, text, context);
        this.checked=checked;
        loadSetting();
    }

    public boolean isChecked(){
        return this.checked;
    }

    public void setChecked(boolean checked){
        this.checked=checked;
        saveSetting();
    }

    public Settings setCheckedChangeListener(CheckedChangeListener changeListener){
        this.checkedChangeListener=changeListener;
        return this;
    }

    public void onCheckedChanged(boolean value){
        setChecked(value);
        if(checkedChangeListener!=null)
            checkedChangeListener.onChangeChecked(checked);
    }

    public interface CheckedChangeListener{
        void onChangeChecked(boolean value);
    }

    @Override
    public void saveSetting(){
        SharedPreferences preferences = getSettings(context);
        @SuppressLint("CommitPrefEdits")
        SharedPreferences.Editor editor = preferences.edit();
        editor.putBoolean(name, checked);
        editor.apply();
    }

    @Override
    public void loadSetting(){
        SharedPreferences preferences = getSettings(context);
        this.checked=preferences.getBoolean(name, checked);
    }
}

File: app/src/main/java/com/dertyp7214/appstore/settings/SettingsSwitch.kt
/*
 * Copyright (c) 2018.
 * Created by Josua Lengwenath
 */

package com.dertyp7214.appstore.settings

import android.annotation.SuppressLint
import android.content.Context
import com.dertyp7214.appstore.Utils.Companion.getSettings

class SettingsSwitch(name: String, text: String, context: Context, private var checked: Boolean) : Settings(name, text, context) {
    private var checkedChangeListener: CheckedChangeListener? = null

    var isChecked: Boolean
        get() = this.checked
        set(checked) {
            this.checked = checked
            saveSetting()
        }

    init {
        loadSetting()
    }

    fun setCheckedChangeListener(changeListener: CheckedChangeListener): SettingsSwitch {
        this.checkedChangeListener = changeListener
        return this
    }

    fun onCheckedChanged(value: Boolean) {
        isChecked = value
        if (checkedChangeListener != null)
            checkedChangeListener!!.onChangeChecked(checked)
    }

    interface CheckedChangeListener {
        fun onChangeChecked(value: Boolean)
    }

    override fun saveSetting() {
        val preferences = getSettings(context)
        @SuppressLint("CommitPrefEdits")
        val editor = preferences.edit()
        editor.putBoolean(name, checked)
        editor.apply()
    }

    override fun loadSetting() {
        val preferences = getSettings(context)
        this.checked = preferences.getBoolean(name, checked)
    }
}
